**Introduction/Business Problem **

Determine in a quantifiable manner where is the best location to open a brewery in the city of Seattle.  As an aspiring brewer I believe the best place to open a brewery is **A)** a location that is part of a tight cluster of oth[](http://)er breweries **B)** of the clusters - the best cluster is the one that has the most food/dining options in close proximity.  The goal is to leverage a data driven approach for identifying the appropriate geographic clusters, ranking them based on proximity/magnitude of food/dining options, and then presenting the output in an easily consumable map.

**Data**

The data for this exercise is going to be captured leveraging the Foursquare API.  Specifically we will be leveraging the **search** endpoint to first find all breweries within the Seattle area.  After this we will then be clustering these breweries based on geography.  Finally we will be using the centerpoints of these clusters to then once again leverage the **search** endpoint to identify the cluster with the greatest preponderance of food/dining options in close proximity.

**Methodology** 

Section which represents the main component of the report where you discuss and describe any exploratory data analysis that you did, any inferential statistical testing that you performed, if any, and what machine learnings were used and why.

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import folium
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
from sklearn.cluster import DBSCAN
import matplotlib.cm as cm
import matplotlib.colors as colors

In [2]:
#set parameters for api call
client_id = 'THIIRDS3XMAMXQHJMABQD0FY3SGAU2VOKO5L5GHZUIMJXP4L' # your Foursquare ID
client_secret = '2SQU0R45T05O0ESS5JCBYIRZCRIPODIXBEG4VOSMM3KZ0LSF' # your Foursquare Secret
version = '20180605' # Foursquare API version

limit = '50'
intent = 'browse'
ne1 = '47.687772, -122.251590'
sw1 = '47.6, -122.430775'

ne2 = '47.599999, -122.251590'
sw2 = '47.531678, -122.430775'

brewery = '50327c8591d4c4b30a586d5d'
food = '4d4b7105d754a06374d81259'

In [3]:
address = 'Seattle, WA'

geolocator = Nominatim(user_agent="sea_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Seattle are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Seattle are 47.6038321, -122.3300624.


In [4]:
url1 = 'https://api.foursquare.com/v2/venues/search?client_id=%s&client_secret=%s&v=%s&intent=%s&ne=%s&sw=%s&categoryId=%s&limit=%s' % (client_id, client_secret, version, intent, ne1, sw1, brewery, limit)
url2 = 'https://api.foursquare.com/v2/venues/search?client_id=%s&client_secret=%s&v=%s&intent=%s&ne=%s&sw=%s&categoryId=%s&limit=%s' % (client_id, client_secret, version, intent, ne2, sw2, brewery, limit)

In [5]:
results1 = requests.get(url1).json()
results2 = requests.get(url2).json()

In [6]:
results1_df = json_normalize(results1['response']['venues'])
results1_df = results1_df[['name', 'location.lat', 'location.lng']]

results2_df = json_normalize(results2['response']['venues'])
results2_df = results2_df[['name', 'location.lat', 'location.lng']]

In [7]:
df = pd.concat([results1_df, results2_df])
df=df.reset_index(drop=True)
df

name  location.lat  location.lng
0           Old Stove Brewing Co - Marketfront     47.609591   -122.343041
1                      Fremont Brewing Company     47.648974   -122.344491
2                     Optimism Brewing Company     47.612816   -122.320571
3                              Redhook Brewlab     47.614222   -122.322761
4                         Pike Brewing Company     47.608161   -122.339923
5                                Stoup Brewing     47.666551   -122.371277
6                               Reuben's Brews     47.665398   -122.373270
7                           Cloudburst Brewing     47.611565   -122.345212
8                      Peddler Brewing Company     47.663757   -122.377057
9                Holy Mountain Brewing Company     47.630795   -122.374597
10  Lagunitas Seattle Taproom & Beer Sanctuary     47.664548   -122.378057
11                     Rooftop Brewing Company     47.655766   -122.373127
12                     Elysian Brewing Company     47.614083   -122.315698
13                    RAM Restaurant & Brewery     47.664122   -122.297880
14                                Obec Brewing     47.666596   -122.372672
15                     Ravenna Brewing Company     47.668255   -122.299271
16                      Lucky Envelope Brewing     47.664948   -122.369146
17                         Jolly Roger Taproom     47.663490   -122.372406
18                    Urban Family Brewing Co.     47.660662   -122.390121
19                   Hale's Ales Brewery & Pub     47.658943   -122.364554
20                 Big Time Brewery & Alehouse     47.657878   -122.313369
21                Burke-Gilman Brewing Company     47.661308   -122.288067
22                            Populuxe Brewing     47.664524   -122.367943
23                         Dirty Couch Brewing     47.661097   -122.391747
24                       McMenamins Queen Anne     47.625399   -122.351690
25                     Floating Bridge Brewing     47.661456   -122.319802
26                     Bad Jimmy's Brewing Co.     47.660078   -122.365553
27                            NW Peaks Brewery     47.664900   -122.378685
28                      Figurehead Brewing Co.     47.656181   -122.383632
29               The Brewtap at Reuben’s Brews     47.662412   -122.366641
..                                         ...           ...           ...
35                  Outer Planet Craft Brewing     47.618050   -122.316821
36                           Urban Beer Garden     47.648977   -122.344414
37                Maritime Pacific Brewing Co.     47.662692   -122.371572
38                                Fremont West     47.663383   -122.367931
39                            Pyramid Alehouse     47.592030   -122.334569
40                              Elysian Fields     47.596988   -122.333508
41                         Pyramid Beer Garden     47.591477   -122.334470
42                          Perihelion Brewery     47.578487   -122.311889
43                    West Seattle Brewing Co.     47.580409   -122.406888
44                 Elliott Bay Brewery and Pub     47.560418   -122.386602
45                            Standard Brewing     47.599396   -122.299711
46                   Jellyfish Brewing Company     47.548936   -122.320457
47                      Locust Cider & Brewing     47.578355   -122.413034
48                     Seapine Brewing Company     47.576040   -122.335427
49                  Georgetown Brewing Company     47.555243   -122.325895
50                       Machine House Brewery     47.550350   -122.317303
51                   Ghostfish Brewing Company     47.576313   -122.333961
52                         Old Rainier Brewery     47.576460   -122.320911
53                           Lowercase Brewing     47.547451   -122.314374
54                         Flying Lion Brewing     47.556077   -122.284319
55                   Best of Hands Barrelhouse     47.535320   -122.376614
56                            Schooner Brewing     47.567842   -122

In [8]:
# create map of Seattle using latitude and longitude values
map_seattle = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, name in zip(df['location.lat'], df['location.lng'], df['name']):
    label = '{}'.format(name)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_seattle)  
    
map_seattle

In [9]:
X = df[['location.lat', 'location.lng']]

In [10]:
epsilon = 0.01
minimumSamples = 4
db = DBSCAN(eps=epsilon, min_samples=minimumSamples).fit(X)
labels = db.labels_
labels

array([-1,  0,  1,  1, -1,  2,  2, -1,  2, -1,  2,  2,  1, -1,  2, -1,  2,
        2,  2,  2, -1, -1,  2,  2, -1, -1,  2,  2,  2,  2,  0, -1, -1,  2,
        0,  1,  0,  2,  2,  3,  3,  3, -1, -1, -1, -1,  4, -1,  5,  4,  4,
        5, -1,  4, -1, -1,  5,  5,  4, -1,  4,  4, -1,  5,  3])

In [11]:
df.insert(0, 'Cluster Labels', labels)

In [12]:
df

Cluster Labels                                        name  location.lat  \
0               -1          Old Stove Brewing Co - Marketfront     47.609591   
1                0                     Fremont Brewing Company     47.648974   
2                1                    Optimism Brewing Company     47.612816   
3                1                             Redhook Brewlab     47.614222   
4               -1                        Pike Brewing Company     47.608161   
5                2                               Stoup Brewing     47.666551   
6                2                              Reuben's Brews     47.665398   
7               -1                          Cloudburst Brewing     47.611565   
8                2                     Peddler Brewing Company     47.663757   
9               -1               Holy Mountain Brewing Company     47.630795   
10               2  Lagunitas Seattle Taproom & Beer Sanctuary     47.664548   
11               2                     Rooftop Brewing Company     47.655766   
12               1                     Elysian Brewing Company     47.614083   
13              -1                    RAM Restaurant & Brewery     47.664122   
14               2                                Obec Brewing     47.666596   
15              -1                     Ravenna Brewing Company     47.668255   
16               2                      Lucky Envelope Brewing     47.664948   
17               2                         Jolly Roger Taproom     47.663490   
18               2                    Urban Family Brewing Co.     47.660662   
19               2                   Hale's Ales Brewery & Pub     47.658943   
20              -1                 Big Time Brewery & Alehouse     47.657878   
21              -1                Burke-Gilman Brewing Company     47.661308   
22               2                            Populuxe Brewing     47.664524   
23               2                         Dirty Couch Brewing     47.661097   
24              -1                       McMenamins Queen Anne     47.625399   
25              -1                     Floating Bridge Brewing     47.661456   
26               2                     Bad Jimmy's Brewing Co.     47.660078   
27               2                            NW Peaks Brewery     47.664900   
28               2                      Figurehead Brewing Co.     47.656181   
29               2               The Brewtap at Reuben’s Brews     47.662412   
..             ...                                         ...           ...   
35               1                  Outer Planet Craft Brewing     47.618050   
36               0                           Urban Beer Garden     47.648977   
37               2                Maritime Pacific Brewing Co.     47.662692   
38               2                                Fremont West     47.663383   
39               3                            Pyramid Alehouse     47.592030   
40               3                              Elysian Fields     47.596988   
41               3                         Pyramid Beer Garden     47.591477   
42              -1                          Perihelion Brewery     47.578487   
43              -1                    West Seattle Brewing Co.     47.580409   
44              -1                 Elliott Bay Brewery and Pub     47.560418   
45              -1                            Standard Brewing     47.599396   
46               4                   Jellyfish Brewing Company     47.548936   
47              -1                      Locust Cider & Brewing     47.578355   
48               5                     Seapine Brewing Company     47.576040   
49               4                  Georgetown Brewing Company     47.555243   
50               4                       Machine House Brewery     47.550350   
51               5                   Ghostfish Brewing Company     47.576313   
52              -1                         Old Rainier Brewery     47.576460   
53  

In [13]:
clusters = df['Cluster Labels'].unique()
clusters

array([-1,  0,  1,  2,  3,  4,  5])

In [14]:
# create map of Seattle using latitude and longitude values
cluster_map = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
#x = np.arange(clusters)
#ys = [i + x + (i*x)**2 for i in range(clusters)]
#colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
#rainbow = [colors.rgb2hex(i) for i in colors_array]

def label_color(label):
    if label == -1:
        return 'grey'
    elif label == 0:
        return 'yellow',
    elif label == 1:
        return 'green',
    elif label == 2:
        return 'blue',
    elif label == 3:
        return 'red',
    elif label == 4:
        return 'pink',
    elif label == 5:
        return 'orange',
    else:
        return 'black'

# add markers to map
markers_colors = []
for lat, lng, name, cluster in zip(df['location.lat'], df['location.lng'], df['name'], df['Cluster Labels']):
    label = folium.Popup(str(name) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color=label_color(cluster),
        fill=True,
        fill_color=label_color(cluster),
        fill_opacity=0.7).add_to(cluster_map)  
    
cluster_map

In [15]:
cluster_df = df.groupby(['Cluster Labels']).mean()
cluster_df

location.lat  location.lng
Cluster Labels                            
-1                 47.608288   -122.338148
 0                 47.650010   -122.344238
 1                 47.614793   -122.318963
 2                 47.662776   -122.374754
 3                 47.592219   -122.332647
 4                 47.550587   -122.321091
 5                 47.568315   -122.336133

In [16]:
radius = '100'

In [17]:
for x in range (0,6):
    record = cluster_df.loc[x]
    lat = record['location.lat']
    long = record['location.lng']
    url3 = 'https://api.foursquare.com/v2/venues/search?client_id=%s&client_secret=%s&v=%s&intent=%s&ll=%s,%s&radius=%s&categoryId=%s&limit=%s' % (client_id, client_secret, version, intent, lat, long, radius, food, limit)
    results3 = requests.get(url3).json()
    results3_df = json_normalize(results3['response']['venues'])
    print (str(x) + ' ' + str(results3_df.shape[0]))

0 0
1 29
2 5
3 24
4 0
5 2


**Results**

Based on the analysis it was clear that there were only a number of defined clusters of breweries within the city of Seattle.  The can be categorized as:
* Cluster 0: Fremont
* Cluster 1: Capitol Hill
* Cluster 2: Ballard
* Cluster 3: Downtown
* Cluster 4: SoDo

While Ballard is clearly the most pronounced cluster in terms of cluster points.  The criteria for stated analysis was to identify clusters and then chose the most optimal location based off of close proximity of food options.  Taking the central point of each cluster and searching against the Foursquare API using a radius of 100 meters Capitol Hill becomes the clear winner.

Thus I would recommend a location at/near the central point of the Capitol Hill Cluster for my new brewery - (47.614793	-122.318963).

**Discussion**

While this finding is interesting there is clearly room for improvement.  Two specific items that stand out are as follows.

DBScan was not fully optimized - and ideally there would be additional analysis done on how the model parameters impacted the outcomes and identification of what the ideal parameters are.

Proximity to Food Locations was not fully vetted.  While 100 meter radius produced a certain result - changing this parameter to greater/lesser values affects the results.

**Conclusion**

By leveraging the DBScan module, identifying clusters of categorical venues, and then assessing those clusters attractiveness based on the cluster's focal points' proximity to other complementary venues (in this case food) we can create an objective means of assessing potential business locations.

This basic framework could be further hardended and parameterized to create a tool for searching for categorical clusters and then assessing favorability in a more user friendly experience/application.  e.g. I want to find clusters of bowling alley's near parks, or so on and so forth - and could become an interesting business location opportunity assessment tool.